### PR TITLE
Expand Newton security guidance

### PR DESCRIPTION
**Pull request by shi-eric**
## Description

Replaces the repository's one-line link to the Newton governance security policy with a Newton-specific security policy. Keeping this context in the repository gives users, maintainers, and automated vulnerability analysis tools the project-specific information needed to assess Newton accurately instead of relying on generic assumptions or an external policy without Newton's implementation details.

The new document covers supported versions, private vulnerability reporting, security boundaries, major threats, critical assumptions, deployment guidance, and dependency and release practices.

The deployment guidance also documents that `ViewerViser` does not expose a host setting and that the pinned Viser release binds its unauthenticated HTTP and WebSocket server to all interfaces by default. This remains true with `share=False`, even though Newton displays a `localhost` URL. The policy recommends firewall or network isolation and an authenticated TLS proxy for intentional remote access.

## Checklist

- [x] New or existing tests cover these changes (not applicable; Markdown-only change)
- [x] The documentation is up to date with these changes
- [x] `CHANGELOG.md` has been updated (not applicable; no behavior change)

## Test plan

- `uvx pre-commit run -a`
- No unit tests were run because this change only updates Markdown documentation.

## Summary by CodeRabbit

* **Documentation**
  * Replaced the security policy with comprehensive guidance on vulnerability reporting and supported releases.
  * Added deployment and integration guidance covering untrusted inputs, remote assets, viewer exposure, checkpoints, and in-process usage.
  * Documented security boundaries, threat categories, dependency responsibilities, reproducible builds, and secure release practices.

### CODE REVIEW

File: SECURITY.md [hunk 1]
```diff
@@ -1 +1,179 @@
...
+The repository's `uv.lock` records resolved versions and artifact hashes for
+development and CI. Published dependency constraints may resolve newer
+versions for downstream users, so consumers should maintain their own tested
+lockfile or equivalent reproducible environment.
+
+Release workflows pin third-party GitHub Actions to commit hashes and use OIDC
+trusted publishing for PyPI. Maintainers should continue reviewing action
+updates, preserving protected release environments, minimizing workflow
+permissions, and separating untrusted pull-request execution from privileged
+release and cloud operations.
```
[the remark this answers is not in the document]
Comment (shi-eric) [reply SECURITY.md]: @coderabbitai This is out of scope. This PR changes only `SECURITY.md` and does not touch `.github/workflows/release.yml`. Please keep comments within the scope of the pull request. Pre-existing workflow permissions should be raised separately, not as a blocking finding on an unrelated documentation change.

File: SECURITY.md [hunk 2]
```diff
@@ -1 +1,179 @@
...
+## Reporting a Vulnerability
+
+If you discover a potential security vulnerability in Newton or one of its
+dependencies, please **do not open a public GitHub issue, pull request, or
+discussion**. Public reports may expose users before a fix is available.
+
+Use the **Security** tab in the relevant repository within the
+[newton-physics organization](https://github.com/newton-physics) and select
+**Report a vulnerability**. For Newton itself, report through the
+[Newton repository's Security tab](https://github.com/newton-physics/newton/security).
```
Comment (jcarius-nv) [inline SECURITY.md]: Could we clarify the reporting path for dependency vulnerabilities? This currently directs findings in “Newton or one of its dependencies” to a relevant repository within the `newton-physics` organization, but dependencies such as Warp, PyTorch, MuJoCo, and OpenUSD are maintained elsewhere. I suggest directing upstream flaws to that dependency’s own private security process, while explaining when a report affecting Newton users should also be submitted through Newton’s private advisory.

File: SECURITY.md [hunk 3]
```diff
@@ -1 +1,179 @@
...
+  imported Python modules are trusted code running with the application
+  process's privileges.
+- Repository access controls, branch protection, GitHub environment
+  protection, OIDC trust policies, and package-registry configuration prevent
+  unauthorized release operations.
+
+## Deployment and Integration Guidance
+
+- Prefer ONNX or current exported PyTorch formats from verified sources. Do not
+  load legacy `.pt` or `.pth` checkpoints received from untrusted users.
```
Comment (jcarius-nv) [inline SECURITY.md]: Could we make explicit that exported `.pt2` files also require a fully trusted source? Newton loads them with `torch.export.load`, and PyTorch documents that API as pickle-based and unsafe for untrusted data. The format preference is useful for compatibility and maintenance, but it should not read as a security boundary; `.pt2`, `.pt`, and `.pth` should all be prohibited for untrusted inputs.

**Review by jcarius-nv — COMMENTED** [review]
Two security-policy accuracy points that should be clarified before merge.

**Comment by jcarius-nv** [comment]
General concern: I think Newton-specific security guidance is valuable, but the proposed document currently mixes two kinds of information with very different maintenance characteristics:

1. **Durable Newton security contracts** — the private reporting path, supported-version policy, Newton's in-process/no-sandbox boundary, the broad asset and extension trust model, and high-impact Newton-specific warnings such as `ViewerViser` network exposure.
2. **Volatile implementation inventory** — the exact model formats and optional integrations currently supported, the present dependency structure, and details of today's CI, cloud-runner, OIDC, environment, and publishing setup.

The first category clearly belongs in `SECURITY.md`. I do not think we want to make the second category part of an authoritative policy that must be kept synchronized whenever dependencies, importers, viewers, checkpoint loaders, or release workflows change. A stale security policy can create false assurance.

Could we keep the durable contracts and Newton-specific warnings, while expressing the volatile pieces in more general terms or linking to their canonical documentation, manifests, or workflows? It would also help to retain a clear distinction between the organization-wide policy owned by `newton-governance` and the Newton-specific threat boundaries owned in this repository.

This is not an objection to adding a Newton-specific policy; it is a request to separate the stable security contract from implementation details we are unlikely to maintain reliably here.

**Comment by jcarius-nv** [comment]
@shi-eric what do you think, is this still something you're up for finishing or should we take over?

**Comment by shi-eric** [comment]
Pushed an update.

### What came out

Removed the inventory parts: the dependency structure paragraph, the supported-format lists, and the point-in-time description of how releases work. `pyproject.toml` and `uv.lock` cover the first two.

### Why the CI, OIDC, and publishing material stayed

I think we should keep it. These are the privileged surfaces in the repo: a publish path to PyPI, `pull_request_target` workflows that run with repository context, and cloud credentials that provision self-hosted runners. Anyone assessing Newton's risk needs to know they exist, and that includes the automated scanners that read this file. Inventory understates the surface once it drifts, which is the false assurance you're describing. A surface description drifts the other way. If we stop using cloud runners, the file is over-cautious rather than wrong.

I rewrote that section as commitments instead of a description of today's setup, and said so in the text: a workflow that stops satisfying one is either a defect to fix or a statement here to correct. One of the commitments is to update this file when the surface changes. If you still think that's too much to keep in sync, I'm fine cutting it back.

### Your two inline comments

For dependencies, there's a new subsection saying to report the flaw through that project's own private security process, since those maintainers own the fix and the advisory. It then lists the three cases where Newton should get a report too, such as when we pin an affected version and users need a coordinated update from us. I also dropped "or one of its dependencies" from the opening line, which was what sent everything to newton-physics in the first place.

For `.pt2`, the policy now says that no serialized format Newton loads is safe for untrusted input. The deployment guidance names `.pt2`, `.pt`, and `.pth` together, cites PyTorch's own warning that `torch.export.load()` uses pickle and should never be pointed at untrusted data, and states that preferring newer formats is about compatibility and maintenance rather than security.

### The governance boundary

Added a Scope section at the top that separates what the governance policy owns from what this file adds.

That policy still says to report a vulnerability in "any of our dependencies" through the org's Security tab, which now contradicts the dependency subsection here. The Scope section notes that Newton's guidance is the more specific one, but it is worth fixing there as well.

File: SECURITY.md [hunk 4]
```diff
@@ -1 +1,280 @@
...
+  checkpoints for its learned-controller integrations. Deserializing these
+  artifacts can reconstruct arbitrary objects and execute code, whichever format
+  is preferred at a given release.
+- **Network boundary:** Asset resolution can issue outbound requests to fetch
+  referenced resources. Optional viewers can open inbound network listeners or
+  connect to remote endpoints.
+- **Compute and native-code boundary:** Warp-generated kernels execute on CPUs
+  or GPUs, and native dependencies and GPU drivers execute with the host
+  process's privileges. Newton does not contain faults originating in those
+  layers.
```
[the remark this answers is not in the document]
Comment (shi-eric) [reply SECURITY.md:135]: @coderabbitai Fixed. I meant "contain" in the containment sense, but you are right that it reads as "does not include," which inverts the claim. Changed to "does not isolate failures originating in those layers" so it matches the wording in the assumptions section.

**Review by jcarius-nv — APPROVED** [review]
Rechecked at head `0625813`. The two security-policy accuracy concerns are addressed: dependency vulnerabilities are routed upstream with clear cases for also notifying Newton, and `.pt2`, `.pt`, and `.pth` are all explicitly trusted-only, with format preference clearly separated from security.

The broader fit concern is also addressed in substance. Volatile dependency and format inventories were removed, the governance/Newton boundary is clear, and the retained CI/OIDC guidance is framed as durable security commitments rather than a point-in-time audit. The exact “Three surfaces” wording is a minor maintenance nit, not a blocker.

All checks are green. Approving.
